### PR TITLE
Fixed #5366: add ID to own-waypoints prefix

### LINKED ISSUE #5366: Waypoint description added automatically
Create a new waypoint (e.g. stage of a multicache) with a description. Create another waypoint without a description (description field is empty). After the waypoint was stored, it has the description of the first waypoint. Edit the description of the second waypoint.  Create a third waypoint without a description. It will get the description of the first waypoint as well. Create another waypoint with a description. It will keep its description.

Master from today morning.

### PR DESCRIPTION
**Pull request by rahu**
Merging new and existing waypoints needs a unique waypoint prefix.

### CODE REVIEW

File: main/src/cgeo/geocaching/utils/Formatter.java [hunk 1]
```diff
@@ -201,7 +201,7 @@ public static String formatWaypointInfo(final Waypoint waypoint) {
         if (waypointType != WaypointType.OWN && waypointType != null) {
             infos.add(waypointType.getL10n());
         }
-        if (Waypoint.PREFIX_OWN.equalsIgnoreCase(waypoint.getPrefix())) {
+        if (waypoint.getPrefix().toUpperCase().startsWith(Waypoint.PREFIX_OWN.toUpperCase())) {
```
Comment (rsudev) [inline main/src/cgeo/geocaching/utils/Formatter.java]: I feel not too comfortable with assuming that getPrefix() never returns null. Could you establish that by adding the necessary NonNull annotations?

File: main/src/cgeo/geocaching/EditWaypointActivity.java [hunk 2]
```diff
@@ -539,6 +539,9 @@ protected Void doInBackground(final Void... params) {
                     finishHandler.sendEmptyMessage(SAVE_ERROR);
                     return null;
                 }
+                if (own) {
+                    waypoint.setPrefix(waypoint.getPrefix()+String.valueOf(cache.getWaypoints().size()));
```
Comment (rsudev) [inline main/src/cgeo/geocaching/EditWaypointActivity.java]: Seems I never finished that comment 4 days ago:
Here I think the prefix is not guaranteed to be unique. If you first add two waypoints, than delete the older one and add another, I assume you end up with a duplicate prefix again.

File: main/src/cgeo/geocaching/models/Waypoint.java [hunk 3]
```diff
@@ -142,11 +142,12 @@ private int order() {
         return cachedOrder;
     }
 
+    @NonNull
     public String getPrefix() {
         return prefix;
     }
 
-    public void setPrefix(final String prefix) {
+    public void setPrefix(final @NonNull String prefix) {
```
Comment (rsudev) [inline main/src/cgeo/geocaching/models/Waypoint.java]: Did I overlook this 4 days ago, or did you amend the commit?

Comment (rahu) [inline main/src/cgeo/geocaching/models/Waypoint.java]: I pushed a new commit fixing this. Did github send a notifiaction about adding a commit changing a existing merge request?

Comment (rsudev) [inline main/src/cgeo/geocaching/models/Waypoint.java]: Not sure if github sent a notification, but looking at the PR again, the ordering of comments and commits also suggest that you changed the commit after my initial comment.
The seem to have changed how they present that, so probably therefore I got confused somewhat.

**Comment by rahu** [comment]
I changed the way to find a unique prefix.

File: main/src/cgeo/geocaching/EditWaypointActivity.java [hunk 4]
```diff
@@ -539,6 +539,20 @@ protected Void doInBackground(final Void... params) {
...
+                            try {
+                                final int suffix = Integer.parseInt(w.getPrefix().substring(3));
+                                if (suffix > max) max = suffix;
+                            } catch(final NumberFormatException ignored) {
+
+                            }
+                        }
+                    }
+                    waypoint.setPrefix(waypoint.getPrefix()+String.valueOf(max+1));
+                }
```
Comment (rsudev) [inline main/src/cgeo/geocaching/EditWaypointActivity.java]: While thinking about this implementation I also scanned our code and cross-checked with gc.com, in order to better understand what we should be doing here.
First of all is the prefix meant to have only to digits or letters, as it replaces the 'GC' part in the geocode when creating a waypoint name, especially for gpx export (PKXXXXX, S1XXXXX) (was not aware of that either).
Therefore already code existed, that was meant to make prefixes unique within the cache (and adhere to the length limit), see GpxSerializer.java,202ff and Geocache.java,1383ff and 1350ff
This was 'damaged' when fixing #4686 but I think the fix to that fix should also adhere to the conceptual limits of the prefix (so cross-checking gpx ex- and import seems a good idea)
On a side-note is there also a gap in this implementation, as you compare the prefix of the other waypoints with "own" (we have a constant for that), but then use the previous prefix to construct the new one (implicitly assuming that "own" was assigned here when constructing the waypoint)

**Comment by rahu** [comment]
I found some time to take a closer look at this. I think the main problem is setting the prefix to "OWN" in EditWaypointActivity. If a prefix is set the existing methods to assing a unique prefix do nothing. So I removed setting the prefix to OWN and use waypoint.isUserDefined instead of testing the prefix starts with "OWN".

At the moment I'm unsure if GpxSerializer lines 230:233 
```
if (StringUtils.isBlank(wp.getPrefix()) || StringUtils.equalsIgnoreCase(Waypoint.PREFIX_OWN, wp.getPrefix())) {
   maxPrefix++;
   wp.setPrefix(StringUtils.leftPad(String.valueOf(maxPrefix), 2, '0'));
}
```
are obsolete.

**Comment by rsudev** [comment]
This version doesn't compile (Jenkins):
```
EditWaypointActivity.java:107: error: cannot find symbol
prefix = waypoint.getPrefix();
```
The prefix variable is needed for the editing case for sure, just initializing it to an empty string (and assigning it while saving!) should do the trick.
WRT the GpxSerializer, I would see it as a fall-back option. There might still be old waypoints around in some databases...

**Comment by rahu** [comment]
Don't know what was wrong yesterday. I would swear I compiled an tested this. :-/ 
However - I fixed the compile error and tested again. Waypoint adding without doubling the note is still working as expected.

**Comment by rsudev** [comment]
Thank you very much for your contribution!

**Pull request merged**